src/Trace/RequestTracer.php

```diff
@@ -0,0 +1,401 @@
+        }
+
+        // If the request was provided with a trace context header, we need to send it back with the response
+        // including whether the request was sampled or not.
+        if ($context->fromHeader()) {
+            $this->persistContextHeader($context);
+        }
+
+        $this->tracer = $context->enabled()
+            ? extension_loaded('stackdriver') ? new ExtensionTracer($context) : new ContextTracer($context)
+            : new NullTracer();
+
+        $spanOptions = $options + [
+            'startTime' => $this->startTimeFromHeaders($headers),
+            'name' => $this->nameFromHeaders($headers),
+            'labels' => []
+        ];
+        $spanOptions['labels'] += $this->labelsFromHeaders($headers);
+        $this->tracer->startSpan($spanOptions);
+
+        register_shutdown_function([$this, 'onExit']);
+    }
+
+    /**
+     * The function registered as the shutdown function. Cleans up the trace and reports using the
+     * provided ReporterInterface. Adds additional labels to the root span detected from the response.
+     */
+    public function onExit()
+    {
+        $responseCode = http_response_code();
+
+        // If a redirect, add the HTTP_REDIRECTED_URL label to the main span
+        if ($responseCode == 301 || $responseCode == 302) {
+            foreach (headers_list() as $header) {
+                if (substr($header, 0, 9) == "Location:") {
+                    $this->tracer->addLabel(self::HTTP_REDIRECTED_URL, substr($header, 10));
+                    break;
+                }
+            }
+        }
+
+        $this->tracer->addLabel(self::HTTP_STATUS_CODE, $responseCode);
+
+        // close all open spans
+        do {
+            $span = $this->tracer->endSpan();
+        } while ($span);
+        $this->reporter->report($this->tracer);
+    }
+
+    /**
+     * Return the tracer used for this request.
+     *
+     * @return TracerInterface
+     */
+    public function tracer()
+    {
+        return $this->tracer;
+    }
+
+    /**
+     * inSpan a callable by creating a TraceSpan that manages the startTime and endTime.
+     * If an exception is thrown while executing the callable, the exception will be caught,
+     * the span will be closed, and the exception will be re-thrown.
+     *
+     * Example:
+     * ```
+     * RequestTracer::inSpan(['name' => 'expensive-operation'], function () {
+     *   // do something expensive
+     * });
+     *
+     * function fib($n) {
+     *   // do something expensive
+     * }
+     * $number = RequestTracer::inSpan(['name' => 'fibonacci'], 'fib', [10]);
+     * ```
+     *
+     * @param array $spanOptions Options for the span.
+     *      {@see Google\Cloud\Trace\TraceSpan::__construct()}
+     * @param  callable $callable    The callable to inSpan.
+     * @return mixed Returns whatever the callable returns
+     */
+    public function _inSpan(array $spanOptions, callable $callable, array $arguments = [])
```

Comment:
I don't like the underscore here, but I can't share a name with the static function. Perhaps I should separate the static/singleton handling from the actual class.